internal/engine/ddl/foreign_procedure.go

```diff
@@ -0,0 +1,340 @@
+        RAISE EXCEPTION 'Non-view procedure "%" called in view-only connection', _procedure;
+    END IF;
+
+    -- 2. if it is owner only, the caller is the owner
+    IF _is_owner_only = TRUE AND _schema_owner != current_setting('ctx.caller') THEN
+        RAISE EXCEPTION 'Procedure "%" is owner-only and cannot be called by user "%" in schema "%"', _procedure, current_setting('ctx.caller'), _dbid;
+    END IF;
+
+    -- 3. it is public
+    IF _is_public = FALSE THEN
+        RAISE EXCEPTION 'Procedure "%" is not public and cannot be foreign called', _procedure;
+    END IF;
+
+    -- 4. our input types match the expected input types
+    IF array_length(_expected_input_types, 1) != 1 THEN
+        RAISE EXCEPTION 'Procedure "%" expects exactly one input type, but got %', _procedure, array_length(_expected_input_types, 1);
+    END IF;
+
+    -- since _arg1 is text, we want to ensure that the input type is text
+    IF _expected_input_types[1] != 'TEXT' THEN
+        RAISE EXCEPTION 'Procedure "%" expects input type "TEXT", but got "%"', _procedure, _expected_input_types[1];
+    END IF;
+
+    IF _expected_input_types[2] != 'INT' THEN
+        RAISE EXCEPTION 'Procedure "%" expects input type "INT", but got "%"', _procedure, _expected_input_types[2];
+    END IF;
+
+    -- 5. our return types match the expected return types
+    IF array_length(_expected_return_types, 1) != 1 THEN
+        RAISE EXCEPTION 'Procedure "%" expects exactly one return type, but got %', _procedure, array_length(_expected_return_types, 1);
+    END IF;
+
+    -- since _out_1 is text, we want to ensure that the return type is text
+    IF _expected_return_types[1] != 'TEXT' THEN
+        RAISE EXCEPTION 'Procedure "%" expects return type "TEXT", but got "%"', _procedure, _expected_return_types[1];
+    END IF;
+
+    -- we now call the procedure. we prefix ds_ to the dbid, as per the Kwil rules.
+    EXECUTE format('SELECT * FROM ds_%I.%I($1, $2)', _dbid, _procedure) INTO _out_1 USING _arg1, _arg2;
+
+	-- or, to return a table, we can do:
+	-- RETURN QUERY EXECUTE format('SELECT * FROM ds_%I.%I($1, $2)', _dbid, _procedure) USING _arg1, _arg2;
+
+END;
+$$ LANGUAGE plpgsql;
+*/
+
+// This is implicitly
+// coupled to the schema defined in internal/engine.execution/queries.go, and therefore is implicitly
+// a circular dependency. I am unsure how to resolve this, but am punting on it for now since the structure
+// of the new parts of the engine are still in flux.
+
+// GenerateForeignProcedure generates a plpgsql function that allows the schema to dynamically
+// call procedures in other schemas, expecting certain inputs and return values. It will prefix
+// the generated function with _fp_ (for "foreign procedure").
+func GenerateForeignProcedure(proc *types.ForeignProcedure) (string, error) {
+	str := strings.Builder{}
+
+	// first write the header
+	str.WriteString(fmt.Sprintf(`CREATE OR REPLACE FUNCTION _fp_%s(_dbid TEXT, _procedure TEXT`, proc.Name))
```

Comment:
I think this function is not a hot path, called only on deploy, but for simple string concatenation it's often a good idea to go with `+`.  The `fmt` printers are relatively heavy duty, using lots of reflection and generally useless operations for simple tasks like this.

```go
str.WriteString(`CREATE OR REPLACE FUNCTION _fp_` + proc.Name + `(_dbid TEXT, _procedure TEXT`)
```

https://dev.to/pmalhaire/concatenate-strings-in-golang-a-quick-benchmark-4ahh

Again, probably not a worry here, but something to be aware of in code that is called a lot.

Reply:
Ah thats good to know. None of the string formatting are hot paths as you mentioned, but I will note that for the future.

I probably won't bother to change this right now, since I actually think we will want to re-evaluate how procedures work under the hood after this release. translating to plpgsql is very hard, and extremely error prone. It will be near impossible to add gas to; I think it will require making it a postgres extension after this initial version.